File: clients/apple/MultiLiveTV/Services/HomeFeed.swift
import Foundation

enum HomeFeed {
    static let fetchSize = 50
    static let initialDisplay = 30
    static let scrollLoadSize = 20
    static let hotSize = 6
    static let recentSize = 6
    static let allRowSize = 6

    enum ContentPhase: Equatable {
        case loading
        case error(String)
        case empty
        case content
    }

    enum LoadMoreAction: Equatable {
        case reveal(displayCount: Int)
        case fetch(page: Int)
        case idle
    }

    static func normalizeTitle(_ name: String) -> String {
        name.trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .replacingOccurrences(of: "\\s+", with: "", options: .regularExpression)
            .replacingOccurrences(of: "[·・:：\\-—_]", with: "", options: .regularExpression)
    }

    static func normalizeYear(_ year: String?) -> String {
        guard let year else { return "" }
        let trimmed = year.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let match = trimmed.range(of: #"\d{4}"#, options: .regularExpression) else {
            return ""
        }
        let digits = String(trimmed[match])
        guard let value = Int(digits), isDisplayableReleaseYear(value) else {
            return ""
        }
        return digits
    }

    /// Years used for ranking. Future placeholders (e.g. 2030) map to the current
    /// calendar year so recently-added titles still surface with this year's films.
    static func sortYearValue(_ year: String?, vodTime: Int = 0, now: Date = Date()) -> Int {
        let current = Calendar(identifier: .gregorian).component(.year, from: now)
        if let digits = rawYearDigits(year), let value = Int(digits) {
            if value >= 1900 && value <= current + 1 {
                return value
            }
            if value > current + 1 {
                return current
            }
        }
        return yearFromVodTime(vodTime, fallback: 0)
    }

    static func yearValue(_ year: String?) -> Int {
        Int(normalizeYear(year)) ?? 0
    }

    /// MacCMS often fills placeholder years (e.g. 2030). Hide those on the card.
    static func isDisplayableReleaseYear(_ year: Int, now: Date = Date()) -> Bool {
        let current = Calendar(identifier: .gregorian).component(.year, from: now)
        return year >= 1900 && year <= current + 1
    }

    private static func rawYearDigits(_ year: String?) -> String? {
        guard let year else { return nil }
        let trimmed = year.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let match = trimmed.range(of: #"\d{4}"#, options: .regularExpression) else {
            return nil
        }
        return String(trimmed[match])
    }

    private static func yearFromVodTime(_ vodTime: Int, fallback: Int) -> Int {
        guard vodTime > 0 else { return fallback }
        let date = Date(timeIntervalSince1970: TimeInterval(vodTime))
        return Calendar(identifier: .gregorian).component(.year, from: date)
    }

    /// Release year first (placeholders treated as this year), then update time.
    static func sortByUpdatedDesc(_ items: [VodItem]) -> [VodItem] {
        items.enumerated()
            .sorted { lhs, rhs in
                let ly = sortYearValue(lhs.element.vodYear, vodTime: lhs.element.vodTime)
                let ry = sortYearValue(rhs.element.vodYear, vodTime: rhs.element.vodTime)
                if ly != ry {
                    return ly > ry
                }
                if lhs.element.vodTime != rhs.element.vodTime {
                    return lhs.element.vodTime > rhs.element.vodTime
                }
                return lhs.offset < rhs.offset
            }
            .map(\.element)
    }

    static func mergeKey(for item: VodItem) -> String {
        mergeKey(title: item.vodName, year: item.vodYear)
    }

    static func mergeKey(title: String, year: String?) -> String {
        let normalized = normalizeTitle(title)
        let year = normalizeYear(year)
        if year.isEmpty { return normalized }
        return "\(normalized)|\(year)"
    }

    /// Prefer an existing title|year key when the new row has no year (or vice versa).
    static func resolvePoolMergeKey(for item: VodItem, existingKeys: [String]) -> String {
        let title = normalizeTitle(item.vodName)
        guard !title.isEmpty else { return mergeKey(for: item) }
        let year = normalizeYear(item.vodYear)
        let bare = title
        let yeared = year.isEmpty ? nil : "\(title)|\(year)"
        let titleKeys = existingKeys.filter { $0 == bare || $0.hasPrefix("\(title)|") }

        if let yeared {
            if titleKeys.contains(yeared) { return yeared }
            if titleKeys.contains(bare) { return yeared }
            return yeared
        }

        let yearedKeys = titleKeys.filter { $0.contains("|") }
        if yearedKeys.count == 1 { return yearedKeys[0] }
        if titleKeys.contains(bare) { return bare }
        return bare
    }

    static func initialDisplayCount(poolLength: Int) -> Int {
        min(initialDisplay, poolLength)
    }

    static func nextDisplayCount(current: Int, poolLength: Int) -> Int {
        min(current + scrollLoadSize, poolLength)
    }

    static func canLoadMore(displayCount: Int, poolLength: Int, apiPage: Int, pageCount: Int) -> Bool {
        displayCount < poolLength || apiPage < pageCount
    }

    static func canLoadMorePages(apiPage: Int, pageCount: Int) -> Bool {
        apiPage < pageCount
    }

    static func hotItems(_ pool: [VodItem]) -> [VodItem] {
        Array(sortedByUpdatedDesc(pool).prefix(hotSize))
    }

    static func recentItems(_ pool: [VodItem]) -> [VodItem] {
        Array(sortedByUpdatedDesc(pool).dropFirst(hotSize).prefix(recentSize))
    }

    static func allItems(_ pool: [VodItem]) -> [VodItem] {
        Array(sortedByUpdatedDesc(pool).dropFirst(hotSize + recentSize))
    }

    static func allRows(items: [VodItem], displayCount: Int) -> [[VodItem]] {
        let visible = Array(items.prefix(max(displayCount, 0)))
        guard !visible.isEmpty else { return [] }
        return stride(from: 0, to: visible.count, by: allRowSize).map { start in
            Array(visible[start..<min(start + allRowSize, visible.count)])
        }
    }

    static func initialAllDisplayCount(allLength: Int) -> Int {
        min(allRowSize, max(allLength, 0))
    }

    static func nextAllDisplayCount(current: Int, allLength: Int) -> Int {
        min(current + allRowSize, max(allLength, 0))
    }

    static func canLoadMoreAll(displayCount: Int, allLength: Int, apiPage: Int, pageCount: Int) -> Bool {
        displayCount < allLength || apiPage < pageCount
    }

    static func nextAllLoadMore(
        displayCount: Int,
        allLength: Int,
        apiPage: Int,
        pageCount: Int,
        isBusy: Bool
    ) -> LoadMoreAction {
        if isBusy { return .idle }
        if displayCount < allLength {
            return .reveal(displayCount: nextAllDisplayCount(current: displayCount, allLength: allLength))
        }
        if apiPage < pageCount {
            return .fetch(page: apiPage + 1)
        }
        return .idle
    }

    static func shouldPrefetchMore(index: Int, itemCount: Int) -> Bool {
        itemCount > 0 && index >= max(itemCount - 2, 0)
    }

    static func nextPageLoadMore(apiPage: Int, pageCount: Int, isBusy: Bool) -> LoadMoreAction {
        if isBusy { return .idle }
        if apiPage < pageCount { return .fetch(page: apiPage + 1) }
        return .idle
    }

    private static func sortedByUpdatedDesc(_ pool: [VodItem]) -> [VodItem] {
        sortByUpdatedDesc(pool)
    }

    static func contentPhase(isLoading: Bool, errorMessage: String?, poolIsEmpty: Bool) -> ContentPhase {
        if isLoading && poolIsEmpty { return .loading }
        if let errorMessage, poolIsEmpty { return .error(errorMessage) }
        if poolIsEmpty { return .empty }
        return .content
    }

    static func nextLoadMore(
        displayCount: Int,
        poolLength: Int,
        apiPage: Int,
        pageCount: Int,
        isBusy: Bool
    ) -> LoadMoreAction {
        if isBusy { return .idle }
        if displayCount < poolLength {
            return .reveal(displayCount: nextDisplayCount(current: displayCount, poolLength: poolLength))
        }
        if apiPage < pageCount {
            return .fetch(page: apiPage + 1)
        }
        return .idle
    }

    static func loadMoreFooterID(displayCount: Int, poolLength: Int, apiPage: Int) -> String {
        "\(displayCount)-\(poolLength)-\(apiPage)"
    }

    struct CategorySwitchPlan: Equatable {
        var showLoading: Bool
        var preserveLoadedPages: Bool
    }

    static func categorySwitchPlan(hasCachedSnapshot: Bool) -> CategorySwitchPlan {
        CategorySwitchPlan(showLoading: !hasCachedSnapshot, preserveLoadedPages: hasCachedSnapshot)
    }

    static func pullRefreshPlan(hasContent: Bool) -> CategorySwitchPlan {
        CategorySwitchPlan(showLoading: !hasContent, preserveLoadedPages: hasContent)
    }

    static func clampedDisplayCount(current: Int, poolLength: Int) -> Int {
        min(max(current, 0), max(poolLength, 0))
    }

    static func replaceFirstPage(pool _: [VodItem], incoming: [VodItem]) -> [VodItem] {
        mergeIntoPool(pool: [], incoming: incoming, isFirstBatch: true)
    }

    static func refreshFirstPage(pool: [VodItem], incoming: [VodItem]) -> [VodItem] {
        let head = replaceFirstPage(pool: [], incoming: incoming)
        let headKeys = Set(head.map { mergeKey(for: $0) })
        return head + pool.filter { !headKeys.contains(mergeKey(for: $0)) }
    }

    static func mergeIntoPool(pool: [VodItem], incoming: [VodItem], isFirstBatch: Bool) -> [VodItem] {
        var orderedKeys: [String] = []
        var map: [String: VodItem] = [:]

        func consume(_ items: [VodItem], overwrite: Bool) {
            for item in items {
                let key = resolvePoolMergeKey(for: item, existingKeys: orderedKeys)
                let bare = normalizeTitle(item.vodName)
                if key != bare, map[bare] != nil {
                    if map[key] == nil {
                        map[key] = map[bare]
                        if !orderedKeys.contains(key) {
                            orderedKeys.append(key)
                        }
                    }
                    map[bare] = nil
                    orderedKeys.removeAll { $0 == bare }
                }
                if map[key] == nil {
                    orderedKeys.append(key)
                    map[key] = item
                } else if overwrite {
                    map[key] = item
                }
            }
        }

        if isFirstBatch {
            consume(pool, overwrite: true)
            consume(incoming, overwrite: true)
            let merged = orderedKeys.compactMap { map[$0] }
            return Array(sortByUpdatedDesc(merged).prefix(fetchSize))
        }

        consume(pool, overwrite: false)
        consume(incoming, overwrite: false)
        var seen = Set<String>()
        var combined: [VodItem] = []
        for item in pool {
            let key = resolvePoolMergeKey(for: item, existingKeys: orderedKeys)
            guard seen.insert(key).inserted else { continue }
            combined.append(map[key] ?? item)
        }
        for key in orderedKeys where !seen.contains(key) {
            guard let item = map[key] else { continue }
            guard seen.insert(key).inserted else { continue }
            combined.append(item)
            if combined.count >= pool.count + fetchSize { break }
        }
        return sortByUpdatedDesc(combined)
    }
}

struct HomeFeedSnapshot: Equatable {
    var pool: [VodItem]
    var displayCount: Int
    var apiPage: Int
    var pageCount: Int
}

struct HomeFeedCache {
    private var storage: [String: HomeFeedSnapshot] = [:]

    static func key(for slug: String?) -> String {
        slug ?? "all"
    }

    mutating func save(_ snapshot: HomeFeedSnapshot, slug: String?) {
        storage[Self.key(for: slug)] = snapshot
    }

    func snapshot(for slug: String?) -> HomeFeedSnapshot? {
        storage[Self.key(for: slug)]
    }
}

File: clients/android/core/src/main/kotlin/com/heibaimiao/multilivetv/home/HomeFeed.kt
package com.heibaimiao.multilivetv.home

import com.heibaimiao.multilivetv.model.VodItem

object HomeFeed {
    const val FETCH_SIZE = 50
    const val INITIAL_DISPLAY = 30
    const val SCROLL_LOAD_SIZE = 20
    const val HOT_SIZE = 6
    const val RECENT_SIZE = 6
    const val ALL_ROW_SIZE = 6

    sealed class ContentPhase {
        data object Loading : ContentPhase()
        data class Error(val message: String) : ContentPhase()
        data object Empty : ContentPhase()
        data object Content : ContentPhase()
    }

    sealed class LoadMoreAction {
        data class Reveal(val displayCount: Int) : LoadMoreAction()
        data class Fetch(val page: Int) : LoadMoreAction()
        data object Idle : LoadMoreAction()
    }

    private val trailingYear = Regex("""(?:19|20)\d{2}$""")
    private val trailingYearDisplay = Regex("""[\s\(（]*((?:19|20)\d{2})[年\)）]*$""")

    fun normalizeTitle(name: String): String =
        name.trim().lowercase()
            .replace(Regex("""\s+"""), "")
            .replace(Regex("""[·・:：\-\—_]"""), "")

    fun titleKeyForMatch(name: String): String =
        normalizeTitle(name).replace(trailingYear, "")

    fun searchKeyword(name: String): String {
        val stripped = name.trim().replace(trailingYearDisplay, "")
        return stripped.ifBlank { name.trim() }
    }

    fun normalizeYear(year: String?): String {
        if (year == null) return ""
        val digits = rawYearDigits(year) ?: return ""
        val value = digits.toIntOrNull() ?: return ""
        return if (isDisplayableReleaseYear(value)) digits else ""
    }

    /**
     * Years used for ranking. Future placeholders (e.g. 2030) map to the current
     * calendar year so recently-added titles still surface with this year's films.
     */
    fun sortYearValue(year: String?, vodTime: Int = 0, nowYear: Int = currentCalendarYear()): Int {
        val digits = rawYearDigits(year)
        if (digits != null) {
            val value = digits.toIntOrNull()
            if (value != null) {
                if (value in 1900..(nowYear + 1)) return value
                if (value > nowYear + 1) return nowYear
            }
        }
        return yearFromVodTime(vodTime, fallback = 0)
    }

    fun yearValue(year: String?): Int = normalizeYear(year).toIntOrNull() ?: 0

    fun isDisplayableReleaseYear(year: Int, nowYear: Int = currentCalendarYear()): Boolean =
        year in 1900..(nowYear + 1)

    fun currentCalendarYear(): Int =
        java.util.Calendar.getInstance().get(java.util.Calendar.YEAR)

    private fun rawYearDigits(year: String?): String? {
        if (year == null) return null
        return Regex("""\d{4}""").find(year.trim())?.value
    }

    private fun yearFromVodTime(vodTime: Int, fallback: Int): Int {
        if (vodTime <= 0) return fallback
        val cal = java.util.Calendar.getInstance()
        cal.timeInMillis = vodTime * 1000L
        return cal.get(java.util.Calendar.YEAR)
    }

    fun sortByUpdatedDesc(items: List<VodItem>): List<VodItem> =
        items.mapIndexed { index, item -> index to item }
            .sortedWith(
                compareByDescending<Pair<Int, VodItem>> {
                    sortYearValue(it.second.vodYear, it.second.vodTime)
                }
                    .thenByDescending { it.second.vodTime }
                    .thenBy { it.first },
            )
            .map { it.second }

    fun mergeKey(item: VodItem): String = mergeKey(item.vodName, item.vodYear)

    fun mergeKey(title: String, year: String?): String {
        val normalized = normalizeTitle(title)
        val yearPart = normalizeYear(year)
        return if (yearPart.isEmpty()) normalized else "$normalized|$yearPart"
    }

    fun resolvePoolMergeKey(item: VodItem, existingKeys: List<String>): String {
        val title = normalizeTitle(item.vodName)
        if (title.isEmpty()) return mergeKey(item)
        val year = normalizeYear(item.vodYear)
        val bare = title
        val yeared = if (year.isEmpty()) null else "$title|$year"
        val titleKeys = existingKeys.filter { it == bare || it.startsWith("$title|") }
        if (yeared != null) {
            if (yeared in titleKeys || bare in titleKeys) return yeared
            return yeared
        }
        val yearedKeys = titleKeys.filter { it.contains("|") }
        if (yearedKeys.size == 1) return yearedKeys[0]
        if (bare in titleKeys) return bare
        return bare
    }

    fun initialDisplayCount(poolLength: Int): Int = minOf(INITIAL_DISPLAY, poolLength)

    fun nextDisplayCount(current: Int, poolLength: Int): Int =
        minOf(current + SCROLL_LOAD_SIZE, poolLength)

    fun contentPhase(isLoading: Boolean, errorMessage: String?, poolIsEmpty: Boolean): ContentPhase = when {
        isLoading && poolIsEmpty -> ContentPhase.Loading
        errorMessage != null && poolIsEmpty -> ContentPhase.Error(errorMessage)
        poolIsEmpty -> ContentPhase.Empty
        else -> ContentPhase.Content
    }

    fun nextLoadMore(
        displayCount: Int,
        poolLength: Int,
        apiPage: Int,
        pageCount: Int,
        isBusy: Boolean,
    ): LoadMoreAction {
        if (isBusy) return LoadMoreAction.Idle
        if (displayCount < poolLength) {
            return LoadMoreAction.Reveal(nextDisplayCount(displayCount, poolLength))
        }
        if (apiPage < pageCount) return LoadMoreAction.Fetch(apiPage + 1)
        return LoadMoreAction.Idle
    }

    fun replaceFirstPage(incoming: List<VodItem>): List<VodItem> =
        mergeIntoPool(emptyList(), incoming, isFirstBatch = true)

    fun mergeIntoPool(pool: List<VodItem>, incoming: List<VodItem>, isFirstBatch: Boolean): List<VodItem> {
        val orderedKeys = mutableListOf<String>()
        val map = mutableMapOf<String, VodItem>()

        fun consume(items: List<VodItem>, overwrite: Boolean) {
            for (item in items) {
                val key = resolvePoolMergeKey(item, orderedKeys)
                val bare = normalizeTitle(item.vodName)
                if (key != bare && map[bare] != null) {
                    if (map[key] == null) {
                        val existing = map[bare] ?: return@consume
                        map[key] = existing
                        if (key !in orderedKeys) orderedKeys += key
                    }
                    map.remove(bare)
                    orderedKeys.removeAll { it == bare }
                }
                if (map[key] == null) {
                    orderedKeys += key
                    map[key] = item
                } else if (overwrite) {
                    map[key] = item
                }
            }
        }

        if (isFirstBatch) {
            consume(pool, overwrite = true)
            consume(incoming, overwrite = true)
            val merged = orderedKeys.mapNotNull { map[it] }
            return sortByUpdatedDesc(merged).take(FETCH_SIZE)
        }

        consume(pool, overwrite = false)
        consume(incoming, overwrite = false)
        val seen = mutableSetOf<String>()
        val combined = mutableListOf<VodItem>()
        for (item in pool) {
            val key = resolvePoolMergeKey(item, orderedKeys)
            if (!seen.add(key)) continue
            combined += map[key] ?: item
        }
        for (key in orderedKeys) {
            if (key in seen) continue
            val item = map[key] ?: continue
            if (!seen.add(key)) continue
            combined += item
            if (combined.size >= pool.size + FETCH_SIZE) break
        }
        return sortByUpdatedDesc(combined)
    }
}

data class HomeFeedSnapshot(
    val pool: List<VodItem>,
    val displayCount: Int,
    val apiPage: Int,
    val pageCount: Int,
)

class HomeFeedCache {
    private val storage = mutableMapOf<String, HomeFeedSnapshot>()

    fun save(snapshot: HomeFeedSnapshot, slug: String?) {
        storage[key(slug)] = snapshot
    }

    fun snapshot(slug: String?): HomeFeedSnapshot? = storage[key(slug)]

    companion object {
        fun key(slug: String?): String = slug ?: "all"
    }
}
